lib/std/array_list.zig

```diff
@@ -45,18 +46,15 @@ pub fn ArrayListAligned(comptime T: type, comptime alignment: ?u29) type {
         /// Initialize with capacity to hold at least num elements.
         /// Deinitialize with `deinit` or use `toOwnedSlice`.
         pub fn initCapacity(allocator: *Allocator, num: usize) !Self {
-            var self = Self.init(allocator);
-
-            const new_memory = try self.allocator.allocAdvanced(T, alignment, num, .at_least);
-            self.items.ptr = new_memory.ptr;
-            self.capacity = new_memory.len;
-
-            return self;
+            var unmanaged = try Unmanaged.initCapacity(allocator, num);
+            return unmanaged.toManaged(allocator);
         }
 
         /// Release all allocated memory.
-        pub fn deinit(self: Self) void {
-            self.allocator.free(self.allocatedSlice());
+        pub fn deinit(self: *Self) void {
```

Comment:
Why the change to `*Self`?

Reply:
To be able to set `self` to `undefined` after deinitializing, which can cause bugs in debug mode. Note that this reflects the other std containers, which also take `*Self` in `deinit()`. `ArrayListAlignedUnmanaged` already did, but `ArrayListAligned` not yet.

Reply:
There are many places where constness needs to be changed though, i'm unsure if i should continue with that.

Reply:
> To be able to set self to undefined after deinitializing,

I see.  I wonder if we should provide a way to do one without the other?  Applications may want to de-allocate the array memory, but may not need or want to "de-initialize" the fields of the array structure.

Here's one example.  If you have a struct that contains an `ArrayListAligned` struct as a field, and it also wants to invalid its memory, then it's going to invalidate the same memory twice:
```zig
struct {
    a: ArrayListAligned,
    // more fields
    pub fn deinit(self: *@This()) {
        self.a.deinit();
        // .. deinitialize more fields
        self.* = undefined;  // we deinitialize the same self.a memory again
    }
}
```

In order to handle cases like these, our data structures need to provide a way to "deinit" their resources without invalidating their contents.  I think it's fine to have `deinit` do both by default, but what if we also expose deallocation of resources without invalidation as well, so maybe `deinitNoInvalidate`?

Then we can do:

```zig
struct {
    a: ArrayListAligned,
    // more fields
    pub fn deinit(self: *@This()) {
        self.a.deinitNoInvalidate(); // we invalidate everything at the end
        // .. deinitialize more fields
        self.* = undefined;
    }
}
```

Reply:
P.S. sorry my comments are mostly orthogonal to this PR, maybe I should open an issue for this?

Reply:
> Applications may want to de-allocate the array memory, but may not need or want to "invalidate" the fields of the array structure.
I think that `deinit` should always be destroying, and containers should provide separate methods for deallocating resources, if supported. This is already the case in many standard library containers, for example ArrayList(Unmanaged) have `.shrink` and `HashMap` has `clearAndFree()`. 

> P.S. sorry my comments are mostly orthogonal to this PR, maybe I should open an issue for this?
I think that would be best.

Reply:
Ok here it is: https://github.com/ziglang/zig/issues/6299